backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,358 @@
+			final IQueueProcessor queueProcessor = getAvailableQueueProcessorForWorkPackage(packageProcessorId).orElse(null);
+
+			if (queueProcessor == null)
+			{
+				logger.warn("*** processWorkPackages: No available QueueProcessor found for workPackageId: {}, c_queue_workpackage.c_queue_packageprocessor_id: {}",
+							workPackage.getC_Queue_WorkPackage_ID(), workPackage.getC_Queue_PackageProcessor_ID());
+
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			setupNewCtxForWorkPackage(workPackage, ctx);
+
+			final boolean successfullyHandled = handleWorkPackageProcessing(queueProcessor, workPackage);
+
+			handledAllWorkPackages = handledAllWorkPackages && successfullyHandled;
+		}
+
+		return handledAllWorkPackages;
+	}
+
+	@NonNull
+	private List<I_C_Queue_WorkPackage> pollAndLockWorkPackages(@NonNull final Properties ctx, @NonNull final List<IQueueProcessor> queueProcessors)
+	{
+		final IQuery<I_C_Queue_WorkPackage> queueProcessorWPQueriesAggregator = queryBL.createQueryBuilder(I_C_Queue_WorkPackage.class)
+				//dev-note: workaround to be able to union multiple queries without applying 'limit' and 'order by' to the end result
+				.addEqualsFilter(I_C_Queue_WorkPackage.COLUMNNAME_C_Queue_WorkPackage_ID, -1)
+				.create();
+
+		final QueryLimit numberOfWorkPackagesPerProcessor = QueryLimit.ONE;
+
+		final List<IQuery<I_C_Queue_WorkPackage>> queueProcessorSpecificQueries = queueProcessors
+				.stream()
+				.map(IQueueProcessor::getQueue)
+				.map(queue -> queue.createQuery(ctx, numberOfWorkPackagesPerProcessor))
+				.filter(Optional::isPresent)
+				.map(Optional::get)
+				.map(lockManager::addNotLockedClause)
+				.collect(ImmutableList.toImmutableList());
+
+		if (queueProcessorSpecificQueries.isEmpty())
+		{
+			return ImmutableList.of();
+		}
+
+		queueProcessorWPQueriesAggregator.addUnions(queueProcessorSpecificQueries, true);
+
+		return lockManager.retrieveAndLockMultipleRecords(queueProcessorWPQueriesAggregator, I_C_Queue_WorkPackage.class);
+	}
+
+	@NonNull
+	private Optional<IQueueProcessor> getAvailableQueueProcessorForWorkPackage(@NonNull final QueuePackageProcessorId packageProcessorId)
+	{
+		return queueProcessors
+				.values()
+				.stream()
+				.filter(processor -> processor.getAssignedPackageProcessorIds().contains(packageProcessorId))
+				.filter(IQueueProcessor::isAvailableToWork)
+				.findFirst();
+	}
+
+	private void validateQueueProcessorAssignment(@NonNull final IQueueProcessor queueProcessor)
+	{
+		if (!getSupportedQueueProcessors().contains(queueProcessor.getClass()))
+		{
+			final String supportedClassNames = getSupportedQueueProcessors().stream().map(Class::getName).collect(Collectors.joining(","));
+
+			throw new AdempiereException("QueueProcessorPlanner can only take: [" + supportedClassNames + "]! but received: " + queueProcessor.getClass().getName());
+		}
+	}
+
+	@NonNull
+	private Integer getPollInterval()
```

Comment:
minor: consider returning `int` and rename the method as `getPollIntervalMillis()`
or, the second option would be to return `Duration`.